### PR TITLE
restructuring for use of more than one database adapter / adapter

### PR DESCRIPTION
**Pull request by joelmdesouza**
-create Adapter interface
-changed postgres to implement the Adapter interface
-modified the tests to be compatible with the changes

### CODE REVIEW

File: postgres/postgres.go [hunk 1]
```diff
@@ -62,6 +67,11 @@ func (s *Stmt) Prepare(db *sqlx.DB, SQL string) (statement *sql.Stmt, err error)
 	return
 }
 
+// Load postgres
+func Load() {
+	config.Adapter = Postgres{}
```
Comment (felipeweb) [inline postgres/postgres.go]: please use a pointer of postgres struct as adapter implementation

File: postgres/postgres.go [hunk 2]
```diff
@@ -127,7 +137,7 @@ func chkInvalidIdentifier(identifer ...string) bool {
 }
 
 // WhereByRequest create interface for queries + where
-func WhereByRequest(r *http.Request, initialPlaceholderID int) (whereSyntax string, values []interface{}, err error) {
+func (adapter Postgres) WhereByRequest(r *http.Request, initialPlaceholderID int) (whereSyntax string, values []interface{}, err error) {
```
Comment (felipeweb) [inline postgres/postgres.go]: pointer receiver

File: postgres/postgres.go [hunk 3]
```diff
@@ -206,7 +216,7 @@ func WhereByRequest(r *http.Request, initialPlaceholderID int) (whereSyntax stri
 }
 
 // SetByRequest create a set clause for SQL
-func SetByRequest(r *http.Request, initialPlaceholderID int) (setSyntax string, values []interface{}, err error) {
+func (adapter Postgres) SetByRequest(r *http.Request, initialPlaceholderID int) (setSyntax string, values []interface{}, err error) {
```
Comment (felipeweb) [inline postgres/postgres.go]: pointer receiver

Comment (felipeweb) [reply postgres/postgres.go]: please do the same in others functions

**Review by felipeweb — CHANGES_REQUESTED** [review]

File: postgres/postgres.go (hunk 1 above)
Comment (joelmdesouza) [reply postgres/postgres.go]: Is it to pass Postgres {} as a reference?
```go
config.Adapter = &Postgres{}
```
In the case already with the change of config.
```go
config.PrestConf.Adapter = &Postgres{}
```

Comment (felipeweb) [reply postgres/postgres.go]: nice!!

**Review by felipeweb — APPROVED** [review]
Like you said you need change the function signature to 
```go
func (adapter *Postgres) SetByRequest(r *http.Request, initialPlaceholderID int) (setSyntax string, values []interface{}, err error) {
```  
@joelmdesouza

**Review by felipeweb — CHANGES_REQUESTED** [review]
Waiting changes is my fault to approve this

**Comment by felipeweb** [comment]
@joelmdesouza  could you please solve the conflicts?

**Comment by joelmdesouza** [comment]
How do I resolve conflicts?
I do not have permission from the Resolve Conflicts button.
I saw that scanner was changed, in my change, I just moved the Scanner.
Will I have to pull the current repository and put my changes?

**Comment by lerrua** [comment]
@joelmdesouza yes, you need to update your local branch with our branch master. Basically getting the master changes, merge with your branch, solve the conflicts and push.

**Comment by joelmdesouza** [comment]
Check if I made it right.

**Review by avelino — APPROVED** [review]
LGTM

**Review by marioidival — DISMISSED** [review]

**Comment by marioidival** [comment]
@joelmdesouza 
postgres/postgres.go:72: config.PrestConf.Adapter undefined (type *config.Prest has no field or method Adapter)

**Comment by felipeweb** [comment]
@marioindival all PRs about this need to test manually

Em qui, 28 de set de 2017 às 23:33, Mario Idival <notifications@github.com>
escreveu:

**Comment by avelino** [comment]
@marioidival This PR depends on other PR, need to test in hand

**Comment by lerrua** [comment]
This PR is waiting https://github.com/prest/config/pull/7

**Comment by crgimenes** [comment]
We had to merge manually to resolve conflicts, so we did merge in another PR. but Joel's commits were kept. So we closed this PR

**Pull request closed without merging**